lib/get-component-name-from-filename.js

```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+/**
+ * @param {String} filename
+ * @param {Object} config
+ * @param {String} config.componentNamePattern
+ */
+module.exports = (filename, config) => {
+  const {componentNamePattern} = config;
+
+  if (componentNamePattern.test(filename)) return filename;
+
+  for (let i = 0; i < filename.length; i++) {
+    const part = filename.slice(0, -i);
+    if (componentNamePattern.test(part)) return part;
+  }
```

Comment:
It looks really clunky, but I'm not sure how to determine the implicit component name without adding another regex for that in the config.